_posts/2019-03-05-build-firefox-extension-with-ringcentral-embeddable.md

```diff
@@ -0,0 +1,304 @@
+        "vendors/libphonenumber-js.min.js",
+        "clickToDial.js",
+        "content.js"
+      ],
+      "css": [
+        "clickToDial.css"
+      ]
+    }
+  ],
+  "icons": {
+    "16": "rc16.png",
+    "32": "rc32.png",
+    "48": "rc48.png",
+    "128": "rc128.png"
+  },
+  "background": {
+    "scripts": ["googleClient.js", "background.js"]
+  },
+  "browser_action": {
+    "default_icon": {
+      "16": "rc16.png",
+      "32": "rc32.png"
+    }
+  },
+  "content_security_policy": "script-src 'self' https://ringcentral.github.io/ringcentral-embeddable; object-src 'self'",
+  "manifest_version": 2,
+  "applications": {
+    "gecko": {
+      "id": "integration-embeddable-firefox@ringcentral.com",
+      "strict_min_version": "60.0"
+    }
+  }
+}
+```
+
+Firefox add-ons `manifest.json` file is very similar as chrome extension. We need to add permissions , content , background and content_security_policy . We will inject JS and CSS files into google related website, so we need to add website url into content_scripts.matches.
+
+With `manifest.json`, we can go to add-ons debug page `about:debugging#addons`, and install it into Firefox.
+
+![firefox extension install](https://user-images.githubusercontent.com/7036536/55455933-2b780300-5618-11e9-8db1-2abc664933dc.png)
+
+Remember to reload add-on every time you update codes.
+
+## Inject RingCentral Embeddable into web page
+
+We will use content_scripts to inject RingCentral Embeddable into web page.
+
+### Inject RingCentral Embeddable with adapter.js
+
+Add following codes into `content.js`
+
+```js
+(function() {
+  var rcs = document.createElement("script");
+  rcs.src = "https://ringcentral.github.io/ringcentral-embeddable/adapter.js";
+  var rcs0 = document.getElementsByTagName("script")[0];
+  rcs0.parentNode.insertBefore(rcs, rcs0);
+})();
+```
+
+### Init widget message listener
+
+Add following codes into `content.js`
+
+```js
+// Listen message from RingCentral Embeddable widget and response:
+window.addEventListener('message', (e) => {
+  const request = e.data;
+  if (!request || !request.type) {
+    return;
+  }
+  if (request.type === 'rc-adapter-pushAdapterState' && !this._registered) {
+    this._registered = true;
+    // To get third party service info from background and register service into widget
+    chrome.runtime.sendMessage({ type: 'rc-register-service' }, (response) => {
+      this.postMessageToWidget({
+        type: 'rc-adapter-register-third-party-service',
+        service: response.service,
+      })
+    });
+    return;
+  }
+  // send widget message to background
+  chrome.runtime.sendMessage(request, (response) => {
+    console.log('response:', response);
+    if (request.type === 'rc-post-message-request') {
+      // send background response message to widget
+      this.responseMessageToWidget(request, response)
+    }
+  });
+});
+```
+
+We will pass request message from widget to extension background, and background will handle and response.
```

Comment:
request message => request `messages`
handle and response => handle and `respond`